File: e2e/cypress-javascript/interactions/validators/TasksValidator.js
const { step } = require('@/decorators/step-decorator');
const { MainPage } = require('@/interactions/pages/MainPage');

class TasksValidator {
  constructor() {
    this.mainPage = new MainPage();
    this.hasTask = step('Validate task with title {title} is visible', this._hasTask.bind(this));
    this.hasNoTask = step('Validate task with title {title} is not visible', this._hasNoTask.bind(this));
    this.hasTaskCount = step('Validate task count is {count}', this._hasTaskCount.bind(this));
  }

  _hasTask(title) {
    this.mainPage.taskRowByTitle(title).should('be.visible');
    return this;
  }

  _hasNoTask(title) {
    this.mainPage.taskTableBody.should('not.contain', title);
    return this;
  }

  _hasTaskCount(count) {
    if (count === 0) {
      this.mainPage.taskRows.should('not.exist');
    } else {
      this.mainPage.taskRows.should('have.length', count);
    }
    return this;
  }
}

module.exports = { TasksValidator };

File: e2e/capybara-ruby/interactions/validators/tasks_validator.rb
# frozen_string_literal: true

require_relative '../../decorators/step_decorator'
require_relative '../../support/capybara_helpers'
require_relative '../pages/main_page'

class TasksValidator
  extend StepDecorator::Methods
  include CapybaraHelpers

  def initialize
    @main_page = MainPage.new
  end

  step 'Validate task with title {title} is visible'
  def has_task(title)
    expect(@main_page.task_row_by_title(title)).to be_visible
    self
  end

  step 'Validate task with title {title} is not visible'
  def has_no_task(title)
    expect(@main_page.task_table_body).not_to have_text(title)
    self
  end

  step 'Validate task count is {count}'
  def has_task_count(count)
    if count.zero?
      expect(page).not_to have_css('[data-testid^="task-title-"]')
    else
      expect(@main_page.task_rows.size).to eq(count)
    end
    self
  end
end
